### DISCUSSION TITLE
Adding bleed to a spritesheet - file order with montage

### DISCUSSION THREAD

**Discussion by sandboks** [ImageMagick/ImageMagick#2352] [Help] [answered]
I need a way to automatically add "bleed" around sprites for a game project. I have a spritesheet, "OutsideGrass1.png", consisting of 32x32 tiles, 8 across and 12 down. 

I want each of those tiles to be extended to be 34x34, by copying the pixels across the edges.

This is what I have so far: (using Windows cmd)

`magick convert -crop 32x32 OutsideGrass1.png sprite.png`
`magick convert sprite-*.png -set option:distort:viewport 96x96-32-32 -virtual-pixel Edge -filter point -distort SRT 0 +repage sprite-bleed.png`
`magick convert sprite-bleed-*.png -crop 34x34+31+31 sprite-bleed.png`
`magick montage sprite-bleed-*.png  -tile 8x12 -geometry 34x34+0+0 -background none OutsideGrass1-bleed.png`
`del sprite-*.png`

This works almost perfectly. But, there's a problem with the file order.

The new spritesheet has been created, and bleed has been added successfully. But, the order of the sprites has been jumbled up.

The first 3 lines create 96 "sprite-bleed" images ("sprite-bleed0.png", "sprite-bleed1.png" ... "sprite-bleed95.png"). Naturally, I want the montage command in line 4 to grab them in sequential order. But instead, the order looks random. 

The problem in line 4 seems to be the "sprite-bleed-*.png" part. If I use "sprite-bleed-[0-9].png", the first 10 sprites are correctly used in order. But only those 10.

What I have is based on this:
https://charliejwalter.net/spritesheet-bleed/

This developer used "sprite-bleed-{0..95}.png", which seems to be a Mac-specific way of getting all of those files in order. But that doesn't work when I try it on Windows, and I can't figure out another way.

Can anyone help me correctly grab the sprites in sequential order?

**Comment by snibgo** [comment, marked as answer]
{0..95} is a bash construct. Bash will take the string "sprite-bleed-{0..95}.png" and expand it to 96 space-separated strings. That construct is not available in Windows CMD or BAT. You could use a "for" loop to build your own long environment variable.

When you use "sprite-bleed-*.png" you get the files in alphabetical order. For example, "sprite-bleed-25.png" will come before "sprite-bleed-3.png". The solution is to use leading zeros, so "sprite-bleed-25.png" will come after "sprite-bleed-03.png".

To get leading zeros, write the output to "sprite-bleed-%02d.png" instead of the simple "sprite-bleed.png". If in  BAT script, double the %%.

You "-crop" the sprite-bleed files in a separate convert. You could combine that with the previous convert. 

(Another possibility is to write to and then read from a single TIFF or MIFF file instead of 96 PNG files. This will be faster. Or you could directly pipe the images, in MIFF format, from the convert to the montage.)

**Reply by sandboks** [reply]
Thank you so much for your reply! I was able to get this working.

I went ahead and made a .bat script that takes in command line arguments:

```
@echo off
set str=%1
set str=%str:.=-padding.%
echo [%1] / [%2] / %str%
set tileSize=32
IF NOT [%2] == [] (
SET "var="&for /f "delims=0123456789" %%i in ("%2") do set var=%%i
if defined var (echo %2 NOT numeric) else (
set tileSize=%2
)
)
FOR /F "tokens=* USEBACKQ" %%F IN (`magick identify -format "%%w" %1`) DO (SET w=%%F)
FOR /F "tokens=* USEBACKQ" %%F IN (`magick identify -format "%%h" %1`) DO (SET h=%%F)
set /a remainderW=w%%tileSize
set /a remainderH=h%%tileSize
echo w = %w% h = %h% rem = %remainderW%/%remainderH%
if not %remainderH%==0 (
echo Looks like the tilesize is wrong - exiting
exit /B
)
if not %remainderW%==0 (
echo Looks like the tilesize is wrong - exiting
exit /B
)
set /a tilesX=w/tileSize
set /a tilesY=h/tileSize
set /a tileSizeB=tileSize*3
echo %tileSizeB%
set /a tileSizeC=tileSize+2
set /a tileSizeD=tileSize-1
magick convert -crop %tileSize%x%tileSize% %1 "sprite-%%02d.png"
magick convert sprite-*.png -set option:distort:viewport %tileSizeB%x%tileSizeB%-%tileSize%-%tileSize% -virtual-pixel Edge -filter point -distort SRT 0 +repage -crop %tileSizeC%x%tileSizeC%+%tileSizeD%+%tileSizeD% sprite-bleed-%%02d.png
magick montage sprite-bleed-*.png -tile %tilesX%x%tilesY% -geometry %tileSizeC%x%tileSizeC%+0+0 -background none %str%
del sprite-*.png
```

Usage: script.bat FILENAME TILESIZE (default 32)

It's ugly and could probably be improved in many ways. But, it works! Maybe that'll be of use to anyone else with the same problem.
